Content.Server/NameIdentifier/NameIdentifierSystem.cs

```diff
@@ -0,0 +1,116 @@
+﻿using Content.Shared.GameTicking;
+using Content.Shared.NameIdentifier;
+using Robust.Shared.Prototypes;
+using Robust.Shared.Random;
+
+namespace Content.Server.NameIdentifier;
+
+/// <summary>
+///     Handles unique name identifiers for entities e.g. `monkey (MK-912)`
+/// </summary>
+public class NameIdentifierSystem : EntitySystem
+{
+    [Dependency] private readonly IPrototypeManager _prototypeManager = default!;
+    [Dependency] private readonly IRobustRandom _robustRandom = default!;
+
+    [ViewVariables]
+    public Dictionary<NameIdentifierGroupPrototype, HashSet<int>> CurrentIds = new();
+
+    public override void Initialize()
+    {
+        base.Initialize();
+
+        SubscribeLocalEvent<NameIdentifierComponent, ComponentInit>(OnComponentInit);
+        SubscribeLocalEvent<RoundRestartCleanupEvent>(CleanupIds);
+
+        InitialSetupPrototypes();
+        _prototypeManager.PrototypesReloaded += OnReloadPrototypes;
+    }
+
+    public override void Shutdown()
+    {
+        base.Shutdown();
+
+        _prototypeManager.PrototypesReloaded -= OnReloadPrototypes;
+    }
+
+    /// <summary>
+    ///     Generates a new unique name/suffix for a given entity and adds it to <see cref="CurrentIds"/>
+    ///     but does not set the entity's name.
+    /// </summary>
+    public string GenerateUniqueName(EntityUid uid, NameIdentifierGroupPrototype proto)
+    {
+        var entityName = Name(uid);
+        if (!CurrentIds.TryGetValue(proto, out var set))
+            return entityName;
+
+        if (set.Count == proto.MaxValue + 1)
+        {
+            // Oh jeez. We're outta numbers.
+            return entityName;
+        }
+
+        // This is kind of inefficient with very large amounts of entities but its better than any other method
+        // I could come up with.
+        var randomVal = _robustRandom.Next(0, proto.MaxValue);
+        while (set.Contains(randomVal))
+        {
+            randomVal = _robustRandom.Next(0, proto.MaxValue);
+        }
```

Comment:
the other way i tried was like this

```csharp
var range = Enumerable.Range(0, proto.MaxValue).Except(set);
var randomVal = _robustRandom.Pick(range.ToArray());
```

which is even more horrible and inefficient! so who knows whats best

Reply:
the tradeoff here is that the current method is fast for small amounts of entities in the group (99% of the time) and slow for very large amounts (like, 9800 monkeys or something) but i think thats reasonable and its only slow on spawning, so